northbound/sensorthings/rest.gateway/src/main/java/org/eclipse/sensinact/sensorthings/sensing/rest/impl/HistoryResourceHelper.java

```diff
@@ -0,0 +1,125 @@
+
+import org.eclipse.sensinact.core.snapshot.ResourceSnapshot;
+import org.eclipse.sensinact.core.twin.TimedValue;
+import org.eclipse.sensinact.northbound.session.SensiNactSession;
+import org.eclipse.sensinact.sensorthings.sensing.dto.Observation;
+import org.eclipse.sensinact.sensorthings.sensing.dto.ResultList;
+import org.eclipse.sensinact.sensorthings.sensing.rest.ExpansionSettings;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import jakarta.ws.rs.core.Application;
+import jakarta.ws.rs.core.UriInfo;
+
+/**
+ * Helper class for accessing historical observation data from the history
+ * provider
+ */
+class HistoryResourceHelper {
+
+    private HistoryResourceHelper() {
+    }
+
+    @SuppressWarnings("unchecked")
+    static ResultList<Observation> loadHistoricalObservations(SensiNactSession userSession,
+            Application application, ObjectMapper mapper, UriInfo uriInfo, ExpansionSettings expansions,
+            ResourceSnapshot resourceSnapshot, int localResultLimit) {
+        ResultList<Observation> list = new ResultList<>();
+        list.value = new ArrayList<>();
+        String historyProvider = (String) application.getProperties().get("sensinact.history.provider");
+
+        if (historyProvider != null) {
+            Integer maxResults = (Integer) application.getProperties().get("sensinact.history.result.limit");
+
+            if (localResultLimit > 0) {
+                maxResults = Math.min(localResultLimit, maxResults);
+            }
+
+            Map<String, Object> params = initParameter(resourceSnapshot);
+            // Get count for the full dataset (for pagination metadata)
+            Long count = (Long) userSession.actOnResource(historyProvider, "history", "count", params);
+            list.count = count == null ? null : count > Integer.MAX_VALUE ? Integer.MAX_VALUE : count.intValue();
+
+            int skip = extractSkipParameter(uriInfo, params);
+            int top = extractTopParameter(uriInfo, params, maxResults);
+            extractOrderByParameter(uriInfo, params);
+
+            List<TimedValue<?>> timed;
+            do {
+                params.put("skip", skip);
+
+                timed = (List<TimedValue<?>>) userSession.actOnResource(historyProvider, "history", "rangeFiltered",
+                        params);
+                list.value.addAll(0, DtoMapper.toObservationList(userSession, application, mapper, uriInfo, expansions,
+                        resourceSnapshot, timed));
+                if (timed.isEmpty()) {
+                    break;
+                } else if (timed.size() == 500) {
+                    skip = skip + list.value.size();
+                }
+
+            } while (list.value.size() < count && list.value.size() < top);
+        }
+        return list;
+    }
+
+    private static void extractOrderByParameter(UriInfo uriInfo, Map<String, Object> params) {
+        List<String> orderbyParam = uriInfo.getQueryParameters().get("$orderby");
```

Comment:
How does this work when they're ordering by different values in the observation, or when they're expanding things and ordering?